### ISSUE TITLE
How to best remove C-dependent libraries from RTD environment?

### ISSUE THREAD

**Issue by guydav** [rtfd/readthedocs.org#2549]
## Details

* Project URL: http://minsketch.readthedocs.io/en/latest/index.html
* Build URL (if applicable): N/A
* Read the Docs username (if applicable): guydav

## Expected Result

I'm trying to follow the FAQ for what should be done if I have libraries that depend on C code. Particularly, I'm using bitarray, gmpy2, and numpy. I tried following the instructions, in the FAQ, but I'm not quite sure how do to the following:

> If such libraries are installed via setup.py, you also will need to remove all the C-dependent libraries from your install_requires in the RTD environment.

Is there any documentation / guidance / example project for how to properly configure setup.py to get the dependencies to install during normal installation but not in RTD?

Thank you!

**Comment by cthoyt** [comment]
@guydav also, you have any ideas what to do if C-dependent libraries are unavoidable, such as `readline`, which is giving me a big problem (#2587)

**Comment by mattberjon** [comment]
@guydav and @cthoyt I had the same issue and found some documentation about it. It worked for me, I hope it will work for both of you. As you saw in the [official documentation][l1], it is advised to mock your C libraries by adding few lines of code in your `conf.py`

```python
import sys
from unittest.mock import MagicMock

class Mock(MagicMock):
    @classmethod
    def __getattr__(cls, name):
            return MagicMock()

MOCK_MODULES = ['pygtk', 'gtk', 'gobject', 'argparse', 'numpy', 'pandas']
sys.modules.update((mod_name, Mock()) for mod_name in MOCK_MODULES)
```
What isn't said (maybe the documentation could be improved on this side), is that you can modify the RTD environment by created at the root a YAML file called `readthedocs.yml` as given in the [documentation as well][l2]. This allows you to alter the documentation compilation process. Moreover, in the documentation given by the library [Astropy][l3], you have an idea of what to do. On my side here the steps I did:

- I activated in the administration of my readthedocs project the `Install your project inside a virtualenv using setup.py install`
- Still in the administration I activated the option `Give the virtual environment access to the global site-packages dir`
- I created a `readthedocs.yml` file at the root of my tree and I edited it with the follow

```yaml
requirements_file: requirements_docs.txt

python:
  version: 3

formats:
  - none
```

Basically, I tell the system to only use Sphinx and another package (to compile the Google docstring format) that are in my requirements_docs.txt and to compile everything using python 3 (I'm pretty sure it would work with python 2.7). I pushed that and it worked.

Hope it'll help!

[l1]: http://docs.readthedocs.io/en/latest/faq.html#i-get-import-errors-on-libraries-that-depend-on-c-modules
[l2]: http://docs.readthedocs.io/en/latest/yaml-config.html
[l3]: http://docs.astropy.org/en/stable/development/affiliated-packages.html

**Comment by cthoyt** [comment]
@mattberjon This worked like a charm. I can't thank you enough for the point in the right direction. If you're ever in Germany, I'll be sure to buy you a beer :)
